### PR TITLE
Allow constants to refer statics

### PR DESCRIPTION
**Pull request by spastorino**
r? @oli-obk, now @eddyb will be happy :heart:

### CODE REVIEW

File: src/test/ui/consts/const-fn-not-safe-for-const.rs [hunk 1]
```diff
@@ -18,12 +18,10 @@ static Y: u32 = 0;
 
 const fn get_Y() -> u32 {
     Y
-        //~^ ERROR E0013
```
Comment (oli-obk) [inline src/test/ui/consts/const-fn-not-safe-for-const.rs]: With the `const_fn` feature gate we do not catch this anymore. Stable const fns will still trigger this as shown in https://github.com/rust-lang/rust/blob/56237d75b4271a8a2e0f47d86ea76ebf6d966152/src/test/ui/consts/min_const_fn/min_const_fn.rs#L90

**Comment by Centril** [comment]
I'm not sure why we are doing this change. If we are to do so, I think an RFC would be in order. For the time being, we have an RFC that explicitly states that this should not be allowed.

Also, do we have any test which forbids the following?

```rust
extern {
    static FOO: u8;
}

const X: u8 = unsafe { FOO };
```

**Comment by spastorino** [comment]
@Centril I've just added the test you mentioned, good point ;).

**Comment by Centril** [comment]
Thanks!

Since the main purpose of this work is to simplify compiler internals rather than as a new language feature, I would suggest moving the existing check to HIR. That shouldn't be too difficult I think -- could be its own pass or we could do it in typeck maybe.

**Comment by oli-obk** [comment]
> I'm not sure why we are doing this change. If we are to do so, I think an RFC would be in order.

> For the time being, we have an RFC that explicitly states that this should not be allowed.

We have? I was under the assumption this is just a restriction left over from the old const evaluator. The miri-based const eval can handle this without any problems. Which RFC are you referring to?

**Comment by oli-obk** [comment]
@Centril and I discussed this a bit via chat. Here's the summary:

1. This does not introduce any new *runtime* behaviour, because one could always have used a `static` item instead of a `const` item and it would have worked just as well
2. This does allow array lengths, enum discriminants and `const` items to take references to statics and even dereference them to get a value out.
3. While `const FOO: &AtomicUsize = &AtomicUsize::new(42);` is illegal because the atomic may be arbitrarily duplicated or deduplicated, `const FOO: &AtomicUsize = &SOME_STATIC;` is legal, because the constant referes always to the same named static item's memory
    * this extends even to lazy or extern statics, while you can make a constant refer to a them, you can't actually read from them because the const evaluator catches that at runtime.
4. The only problematic situation that exists is `const fn`. Right now `const fn` cannot refer to `static` items, because that would allow one to write a `const fn` like
    ```rust
    const fn foo() -> usize {
        static FOO: AtomicUsize = AtomicUsize::new(0);
        FOO.fetch_add(1, Ordering::Relaxed)
    }
    ```
    (assuming powerful enough const eval features).
    * Even if we forbade statics, we could use the `const BAR: &AtomicUsize = &FOO;` indirection to work around the static check.
    * While we could just forbid constants in `const fn` that refer to statics anywhere, that won't work for associated constants like
         ```rust
         const fn foo<T: Trait>() -> usize {
             T::ASSOC_CONST.fetch_add(1, Ordering::Relaxed)
         }
         ```
     * We could still do the check at monomorphization time, but that would mean that a use of a function would fail to compile depending on the type the user passed for the generic parameter

cc @RalfJung

**Comment by RalfJung** [comment]
The crucial thing to make sure is that CTFE does not *read from* (let alone write do) mutable statics. We should make sure to have a few miri-unleash tests that cover this. [Cnetril](https://github.com/rust-lang/rust/pull/66302#issuecomment-552762896) mentioned one, @oli-obk mentioned this one:
```rust
const FOO: usize {
    static FOO: AtomicUsize = AtomicUsize::new(0);
    FOO.fetch_add(1, Ordering::Relaxed)
};
```
and then there also is
```rust
static mut MUTABLE: u32 = 0;
const BAD: u32 = unsafe { MUTABLE };
```
All of these should fail even with miri-unleashed, to make sure that our dynamic checks are covering all possibilities.

What remains then is the question of whether we want to have static checks on top of the dynamic ones, and what those are good for. Doing only dynamic checks leads to monomorphization-time errors. But from how I understood @oli-obk, we either already have those or ruling those out also rules out way too many useful things, and hence monomorphization-time errors are the better option? I do not have a strong stanza on this, if @oli-obk and @Centril agree on that that's good enough for me. I just want to see the Miri engine do enough checks to rule out all these examples. :)  (And I think so far it does *not* do enough checks, actually.)

**Comment by JohnCSimon** [comment]
Ping from triage 
@spastorino  Can you please address the merge conflicts?
Thank you!

**Comment by spastorino** [comment]
@JohnCSimon, I don't think it makes sense to rebase this until we don't figure out a way to make this idea work.

**Comment by Centril** [comment]
> But from how I understood @oli-obk, we either already have those or ruling those out also rules out way too many useful things, and hence monomorphization-time errors are the better option? I do not have a strong stanza on this, if @oli-obk and @Centril agree on that that's good enough for me.

My stance is that should not be adding post-monomorphization errors and that `const fn`s should remain deterministic when executed at run-time.

**Comment by oli-obk** [comment]
I'm currently unable to come up with a non-postmonomorphization plan that would not end up making const qualif more complex or require new marker traits

Let's close until we have a plan for working this out

**Comment by RalfJung** [comment]
CTFE already causes post-monomorphization errors because some parts of CTFE can only happen after monomorphization and sometimes that can error (causing a panic during CTFE is the most obvious case).

Not sure if I would consider "you read from a static" a * new* post-monomorphization error.

**Comment by oli-obk** [comment]
writing to mutable statics by first taking a reference and then writing through that reference is also only detected post monomorphization

**Pull request closed without merging**